File: src/boxing_coach/analysis/rule.py
"""The `Rule` seam — the other axis of change the spec calls out.

Building 20-30 of these is where the boxing domain expertise lives. Adding a
rule = one subclass; the engine discovers nothing implicitly and nothing else
changes. Each rule declares which drill focus tags it applies to so the engine
can skip irrelevant rules for a given round.
"""

from __future__ import annotations

from abc import ABC, abstractmethod

from ..domain.analysis import Observation
from .context import AnalysisContext


class Rule(ABC):
    """A single technique check over a round's pose sequence."""

    #: Stable identifier, e.g. "guard_return". Appears on every Observation.
    id: str = ""

    #: Drill focus tags this rule is relevant to. A rule with an empty set is
    #: always run; otherwise it runs only when the drill targets one of them.
    focus_tags: frozenset[str] = frozenset()

    def applies_to(self, context: AnalysisContext) -> bool:
        if not self.focus_tags:
            return True
        return context.drill.is_focused_on(*self.focus_tags)

    @abstractmethod
    def evaluate(self, context: AnalysisContext) -> list[Observation]:
        """Return zero or more observations. Empty means nothing to report."""
        raise NotImplementedError

File: app/lib/analysis/rule.dart
import 'context.dart';
import 'round_analysis.dart';

/// The `Rule` seam — mirror of `src/boxing_coach/analysis/rule.py`.
///
/// A single technique check over a round's pose sequence. Adding a rule is one
/// subclass; the engine discovers nothing implicitly. Each rule declares which
/// drill focus tags it is relevant to so the engine can skip irrelevant rules.
abstract class Rule {
  /// Stable identifier, e.g. "guard_return". Appears on every [Observation].
  String get id;

  /// Drill focus tags this rule is relevant to. Empty = always run; otherwise
  /// it runs only when the drill targets one of them.
  Set<String> get focusTags => const <String>{};

  bool appliesTo(AnalysisContext context) =>
      focusTags.isEmpty || context.drill.isFocusedOn(focusTags);

  /// Zero or more observations. Empty means nothing to report.
  List<Observation> evaluate(AnalysisContext context);
}
